### DISCUSSION TITLE
How to send http request through specific network interface?

### DISCUSSION THREAD

**Discussion by ChrisYP** [encode/httpx#2635] [General] [answered]
I have two network interfaces and both with internet access. Let's say my interfaces are eth0 and ETH999. I need specific requests to go through ETH999 interface and eth0 is the default interface in my server.
I can do this with these codes when I use `requests`.

```
import socket
import requests
from requests import adapters
from urllib3.poolmanager import PoolManager


class InterfaceAdapter(adapters.HTTPAdapter):

    def __init__(self, **kwargs):
        self.iface = kwargs.pop('iface', None)
        super(InterfaceAdapter, self).__init__(**kwargs)

    def _socket_options(self):
        if self.iface is None:
            return []
        else:
            return [(socket.SOL_SOCKET, socket.SO_BINDTODEVICE, self.iface)]

    def init_poolmanager(self, connections, maxsize, block=False):
        self.poolmanager = PoolManager(
            num_pools=connections,
            maxsize=maxsize,
            block=block,
            socket_options=self._socket_options()
        )


session = requests.Session()
for prefix in ('http://', 'https://'):
    # set ETH999 interface here
    session.mount(prefix, InterfaceAdapter(iface=ETH999'))

print(session.get("https://ifconfig.me/").text
```

so how to do the same on Httpx? thanks.

**Comment by lovelydinosaur** [comment]
Hi there. Is this what you're looking for?...

https://www.python-httpx.org/advanced/#custom-transports

> For some advanced configuration you might need to instantiate a transport class directly, and pass it to the client instance. One example is the local_address configuration which is only available via this low-level API.

```python
>>> import httpx
>>> transport = httpx.HTTPTransport(local_address="0.0.0.0")
>>> client = httpx.Client(transport=transport)
```

**Reply by ChrisYP** [reply]
This doesn't work for me. I have two real physical network cards instead of one card with two IP addresses. I need to specify the second network card for network requests. With your settings, I can only get timeout request errors.

**Reply by lovelydinosaur** [reply]
Oh okay, looking at this more carefully now, you need to set the socket options, not the local address, right?

What does your `# set ETH999 interface here` code look like?

**Reply by ChrisYP** [reply]
```
import socket
import requests
from requests_toolbelt.adapters.socket_options import SocketOptionsAdapter

session = requests.Session()
# set interface here  -> ETH999 is my interface's name.
options = [(socket.SOL_SOCKET, socket.SO_BINDTODEVICE, b"ETH999")]
for prefix in ('http://', 'https://'):
    session.mount(prefix, SocketOptionsAdapter(socket_options=options))

print(session.get("https://httpbin.org/ip").text)
```
The above code is work very well for me. Thank you.

**Reply by lovelydinosaur** [reply]
I've opened an issue for this here... https://github.com/encode/httpcore/issues/662

**Reply by lovelydinosaur** [reply]
The `httpcore` package version 0.17.2 now supports this, and I've opened https://github.com/encode/httpx/issues/2715 to add it to `httpx`.

**Comment by karpetrosyan** [comment, marked as answer]
Since #2716 has been merged, hopefully this api will provide the answer to your question beginning with version 0.24.2.

``` python
>>> import httpx
>>> socket_options = [(socket.SOL_SOCKET, socket.SO_BINDTODEVICE, b"ETH999")]
>>> transport = httpx.HTTPTransport(socket_options=socket_options)
>>> client = httpx.Client(transport=transport)
```

**Reply by pinterior** [reply]
Leaving this here for anyone who finds this discussion.

If there are two network interfaces, both with a default route, and trying to use the one with a low-priority (high metric), this does not work.
httpcore's sync backend sets the SO_BINDTODEVICE socket options after connect, so the source address (an address set in high priority interface) has already been selected.
